File: Assets/Scripts/MazeGen/ClusterMaze.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class ClusterMaze
{
    int width, height;
    Cluster[,] grid;
    // Start is called before the first frame update

    public int Length { get { return width * height; } }

    public ClusterMaze(int dimensionX, int dimensionY)
    {
        width = dimensionX;
        height = dimensionY;
        grid = new Cluster[dimensionX, dimensionY];
    }

    public void SetCluster(Coordinate coordinate, Cluster cluster)
    {
        if (grid == null)
            Debug.LogWarning("Trying to set cluster on an empty grid");

        grid[coordinate.X, coordinate.Y] = cluster;
    }
    public bool TryGetCluster(Coordinate coordinate, out Cluster cluster)
    {
        bool outOfBounds = coordinate.X < 0 || coordinate.X > grid.GetLength(0) - 1 || coordinate.Y < 0 || coordinate.Y > grid.GetLength(1) - 1;
        if (outOfBounds)
        {
            cluster = null;
            return false;
        }

        cluster = grid[coordinate.X, coordinate.Y];
        return true;
    }

    public bool CanGetCluster(Coordinate coordinate)
    {
        bool outOfBounds = coordinate.X < 0 || coordinate.X > grid.GetLength(0) - 1 || coordinate.Y < 0 || coordinate.Y > grid.GetLength(1) - 1;
        if (outOfBounds)
        {
            return false;
        }

        return true;
    }

    public int GetLength(int dimension)
    {
        int length = 0;

        switch (dimension)
        {
            case 0:
                length = grid.GetLength(0);
                break;
            case 1:
                length = grid.GetLength(1);
                break;
            default:
                Debug.LogWarning("Invalid dimension requested");
                break;
        }

        return length;
    }

    public List<Cluster> GetLowestEntropy()
    {
        List<Cluster> retval = new List<Cluster>();
        retval.Clear();
        int minEntropy = int.MaxValue;

        // Find tile(s) with lowest entropy (>1)
        for (int x = 0; x < width; x++)
        {
            for (int y = 0; y < height; y++)
            {
                Cluster cluster = grid[x, y];
                int entropy = cluster.Entropy;
                if (entropy > 1)
                {
                    if (entropy < minEntropy)
                    {
                        minEntropy = entropy;
                        retval.Clear();
                        retval.Add(cluster);
                    }
                    else if (entropy == minEntropy)
                    {
                        retval.Add(cluster);
                    }
                }
            }
        }


        return retval;
    }

    public void DestroyMaze()
    {
        grid = null;
    }
    
}

File: lua/clustermaze.lua
ClusterMaze = {}
ClusterMaze.__index = ClusterMaze

function ClusterMaze:new(dimensionX, dimensionY)
    local instance = setmetatable({}, ClusterMaze)
    instance.width = dimensionX
    instance.height = dimensionY
    instance.grid = {}
    for x = 1, dimensionX do
        instance.grid[x] = {}
        for y = 1, dimensionY do
            instance.grid[x][y] = nil
        end
    end
    return instance
end

function ClusterMaze:GetLength()
    return self.width * self.height
end

function ClusterMaze:SetCluster(coordinate, cluster)
    if self.grid == nil then
        print("Warning: Trying to set cluster on an empty grid")
    end
    self.grid[coordinate.x + 1][coordinate.y + 1] = cluster
end

function ClusterMaze:TryGetCluster(coordinate)
    local outOfBounds = coordinate.x < 0 or coordinate.x >= #self.grid or coordinate.y < 0 or coordinate.y >= #self.grid[1]
    if outOfBounds then
        return nil, false
    end
    return self.grid[coordinate.x + 1][coordinate.y + 1], true
end

function ClusterMaze:CanGetCluster(coordinate)
    local outOfBounds = coordinate.x < 0 or coordinate.x >= #self.grid or coordinate.y < 0 or coordinate.y >= #self.grid[1]
    return not outOfBounds
end

function ClusterMaze:GetLengthByDimension(dimension)
    local length = 0
    if dimension == 0 then
        length = #self.grid
    elseif dimension == 1 then
        length = #self.grid[1]
    else
        print("Warning: Invalid dimension requested")
    end
    return length
end

function ClusterMaze:GetLowestEntropy()
    local retval = {}
    local minEntropy = math.huge

    for x = 1, self.width do
        for y = 1, self.height do
            local cluster = self.grid[x][y]
            if cluster and cluster.Entropy > 1 then
                local entropy = cluster.Entropy
                if entropy < minEntropy then
                    minEntropy = entropy
                    retval = {cluster}
                elseif entropy == minEntropy then
                    table.insert(retval, cluster)
                end
            end
        end
    end

    return retval
end

function ClusterMaze:DestroyMaze()
    self.grid = nil
end
